Adding audio file feature & langchain chunker

isyanap94 opened the pull request.

Commit:
Audio data loader initial commit:
- Load audio transcript for RAG by providing url/local audio path
- Chunking is still using default strategy, which is not optimized enough for transcription
- Diarize feature still deactivated (WIP)

Commit:
Audio data loader initial commit:
- Load audio transcript for RAG by providing url/local audio path
- Chunking is still using default strategy, which is not optimized enough for transcription
- Diarize feature still deactivated (WIP)

Commit:
Add new test configs

Commit:
Add new test configs

Commit:
- Add char removal except comma & period
- Import new class & change the method name

Commit:
- Add char removal except comma & period
- Import new class & change the method name

Commit:
Move the text clean on text parse method

Commit:
Move the text clean on text parse method

Commit:
- Add new chunking method using langchain with chunk_overlap
- Add server health check before sending prompt

Commit:
- Add new chunking method using langchain with chunk_overlap
- Add server health check before sending prompt

Commit:
Add new param & remove overlap chunking

Commit:
Add new param & remove overlap chunking

Commit:
Move default value to config

Commit:
Move default value to config

dmb23 commented on README.md:
```diff
@@ -14,9 +14,15 @@ Again, mostly meant for experimentation. But allows to use via a CLI when instal
-- download a model
+- download the models and put it under /dragtor/models folder
     - I use [a quantized Llama3.1 8B Instruct](https://huggingface.co/bartowski/Meta-Llama-3.1-8B-Instruct-GGUF/blob/main/Meta-Llama-3.1-8B-Instruct-Q4_K_M.gguf)
+    - For diarize (feature is not yet active), use [base EN model](https://huggingface.co/ggerganov/whisper.cpp/blob/main/ggml-base.en.bin)
```
if diarization is not implemented, then leave it out

dmb23 commented on README.md:
```diff
@@ -14,9 +14,15 @@ Again, mostly meant for experimentation. But allows to use via a CLI when instal
     - I use [a quantized Llama3.1 8B Instruct](https://huggingface.co/bartowski/Meta-Llama-3.1-8B-Instruct-GGUF/blob/main/Meta-Llama-3.1-8B-Instruct-Q4_K_M.gguf)
+    - For diarize (feature is not yet active), use [base EN model](https://huggingface.co/ggerganov/whisper.cpp/blob/main/ggml-base.en.bin)
 - `uv run dragtor` to start via the CLI
+- install PyTorch, to use Mac-optimized packages:
```
I would assume you don't need torch if you don't use diarization

dmb23 commented on config/params.yml:
```diff
@@ -59,3 +68,11 @@ prompts:
+  files: "/Users/mischa/Projects/local/dragtor/data/audio_files/"
+  diarization_model: "pyannote/speaker-diarization-3.1"
+  lang: "en"
+  num_speakers: 1
```
that seems like a bad default?

dmb23 commented on pyproject.toml:
```diff
@@ -3,13 +3,15 @@ name = "dragtor"
     "loguru>=0.7.2",
     "fire>=0.6.0",
     "sentence-transformers>=3.0.1",
+    "pyannote.audio>=3.3.2",
```
again, is this needed currently?

dmb23 commented on src/dragtor/audio_loader.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        # Set default values from config if not provided
+        language = language or config.conf.audio.lang
+        num_speakers = num_speakers or config.conf.audio.num_speakers
```
Ideally: either via config (and / or cli flag) or via argument

dmb23 commented on src/dragtor/audio_loader.py:
```diff
@@ -0,0 +1,108 @@
+from dragtor.utils.audio_utils import *
```
Please don't use star imports!

dmb23 commented on src/dragtor/utils/audio_diarization_utils.py:
```diff
@@ -0,0 +1,224 @@
+import re
```
I feel like all of these functions should be private methods of the `AudioLoader` class - they are not meant to be used on their own or in multiple different contexts

dmb23 commented on src/dragtor/utils/audio_utils.py:
```diff
@@ -0,0 +1,67 @@
+    return audio_path
+
+
+def parse_transcript(transcript: str) -> str:
```
extremely unclear what is the difference between `parse_transcript`, `parse_audio_transcript` and `parse_diarization`

dmb23 commented on src/dragtor/utils/audio_utils.py:
```diff
@@ -0,0 +1,67 @@
+    logger.info(f"Processing: {audio_path}")
+    if audio_path.startswith(("http://", "https://")):
+        inputs = requests.get(audio_path).content
+        file_path = output_dir / "_".join(Path(audio_path).parts[-2:])
```
this will break!

dmb23 commented on src/dragtor/utils/audio_utils.py:
```diff
@@ -0,0 +1,67 @@
+    return transcript_segments
+
+
+def cleanup():
```
if you are working with temporary files for processing: look into https://docs.python.org/3/library/tempfile.html

dmb23 commented on tests/test_audio_data.py:
```diff
@@ -0,0 +1,89 @@
+
+    monkeypatch.setattr("subprocess.run", mock_run)
+
+# Test for preprocess_audio_file function with a URL
```
- is this supposed to be a docstring for the function or a TODO?

dmb23 commented on tests/test_audio_data.py:
```diff
@@ -0,0 +1,89 @@
+from os.path import isfile
```
In these tests you are mostly testing your own mocks...

Commit:
Updates for resolving feedbacks

Commit:
Updates for resolving feedbacks

Commit:
Some changes:
- Change audio loader to use temp file & dir
- Remove diarize logic (stored in different branch)
- Remove utils files

Commit:
Update logic on AudioLoader

Commit:
Merge remote-tracking branch 'origin/feat/audio-feature' into feat/audio-feature

# Conflicts:
#	README.md
#	config/params.yml
#	pyproject.toml
#	src/dragtor/audio_loader.py
#	src/dragtor/cli.py
#	src/dragtor/llm/llm.py

Commit:
Remove audio utils

Commit:
Rebase

Commit:
Rebase

Commit:
Update logic for testing

Commit:
Add local audio for testing purpose

Commit:
Small updates

dmb23 commented on config/params.yml:
```diff
@@ -1,6 +1,8 @@
 # allows to use loguru for external loggers
 expose_library_logs: False
+# set GPU accelerator. mps for Apple ecosystem, CUDA for NVIDIA, etc.
+device: "mps"
```
I think this is unused?

dmb23 commented on pyproject.toml:
please update also the uv.lock file when you change the requirements

dmb23 commented on tests/test_audio_data.py:
```diff
@@ -0,0 +1,63 @@
+from dragtor.audio_loader import AudioLoader
+from dragtor import config
+
+project_root = Path(__file__).parent.parent
+local_audio_sample = project_root / "tests" / "assets" / "audio_sample" / "sample.wav"
+local_invalid_audio_sample = project_root / "tests" / "assets" / "audio_sample" / "sample99.mp3"
```
you can have a look in conftest.py how you can leverage importlib.resources for file handling in tests

dmb23 commented on tests/test_audio_data.py:
these tests fail for me (except the ffmpeg availability). Can you please check if these work for you in a clean setup?

isyanap94 replied:
Now it works after some minor adjustment. The readme files missing the guide for download the transcript model

dmb23 commented on src/dragtor/cli.py:
test_cli_load fails for me. Can you also validate that?

dmb23 commented on src/dragtor/index/chunk.py:
```diff
@@ -68,13 +69,46 @@ def chunk_and_annotate(self, text: str) -> tuple[list[str], list[tuple[int, int]
 
         return chunks, span_annotations
 
+class LangchainChunker(Chunker):
```
Langchain provides multiple different chunkers, so this name is misleading. I would prefer something like `RecursiveCharacterChunker`

dmb23 reviewed (commented):
Audio loader looks nice. But please check the tests (if they run for you, check in a clean setup with fresh clone and new uv environment)

Commit:
Update chunker name & test module

Commit:
Python var adjustment & minor change in test modules

Commit:
Minor update

Commit:
Add test for new chunker

Commit:
Minor change

dmb23 commented on src/dragtor/audio_loader.py:
```diff
@@ -0,0 +1,165 @@
+
+    def _transcribe_audio(self, wav_file) -> str:
+        """Transcribe .wav audio file using transcription model."""
+        command = f"transcribe -m '{self.model_path}' -f '{str(wav_file)}' -l '{self.language}'"
```
`transcribe` is not an executable from building whisper.cpp - where does it come from?

dmb23 commented on src/dragtor/audio_loader.py:
```diff
@@ -0,0 +1,165 @@
+        process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, error = process.communicate()
+        if process.returncode != 0:
+            raise RuntimeError(f"Transcription process failed for {wav_file}. Make sure transcription model is available under models/")
```
you lose any information about why the transcription process failed here - might be helpful to log the error from the subprocess.

dmb23 commented on transcribe:
Please remove this file. It was compiled for your machine and will break almost everywhere else. Instead give instructions in the README to install whisper.cpp and what needs to happen so that dragtor can call the correct executables!

Commit:
Add steps for compiling whisper.cpp & use importlib.resources

Commit:
Merge remote-tracking branch 'origin/main' into feat/audio-feature

# Conflicts:
#	config/params.yml

Commit:
Merge params.yml recent change

Commit:
Remove exe file

Commit:
Set the same test url

Commit:
Add one more line

dmb23 commented on README.md:
```diff
@@ -6,17 +6,36 @@ This is a solution to get grounded LLM advice on climbing-related health and tra
+     - Transcription model can also be found under /whisper.cpp/models
+     - You can also directly copy this model into /dragtor/models. Using Mac terminal: `cp models/ggml-base.en.bin /<your-path>/dragtor/models/`
+  4. After a few minutes, find an executable file with name `main`
+  5. Copy it to /dragtor folder, and rename it to `transcribe`
```
if you recommend to copy a file to the dragtor folder, then place it in the .gitignore file

dmb23 commented on README.md:
```diff
@@ -6,17 +6,36 @@ This is a solution to get grounded LLM advice on climbing-related health and tra
+  4. After a few minutes, find an executable file with name `main`
+  5. Copy it to /dragtor folder, and rename it to `transcribe`
+     - If you're using Mac terminal: `cp main /<your-path>/dragtor/transcribe`
+- Add dragtor project path into your system path
```
This should not be necessary if you copy the file to the main directory for most use cases...

dmb23 commented on assets/speech_samples_sample3.txt:
why is this file in the assets folder? It seems to me it should be somewhere in tests/ ?

dmb23 commented on src/dragtor/index/chunk.py:
```diff
@@ -68,13 +69,46 @@ def chunk_and_annotate(self, text: str) -> tuple[list[str], list[tuple[int, int]
             return ParagraphChunker()
         case "jina_tokenizer":
             return JinaTokenizerChunker()
+        case "langchain":
```
adjust to new name?

dmb23 commented on tests/test_audio_data.py:
```diff
@@ -0,0 +1,66 @@
+from dragtor.audio_loader import AudioLoader
+from dragtor import config
+
+def test_ffmpeg_availability():
+    """Test if your local machine already installed with ffmpeg."""
+    try:
+        result = subprocess.run(["ffmpeg", "-version"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        assert result.returncode == 0
+    except FileNotFoundError:
+        pytest.fail("ffmpeg is not installed or not available in the system's PATH")
```
This is a good idea! I think it would be great to have the same checks for 
- other external tools (llama-server + whisper)
- check that the defined model files are actually present

If you don't want to implement this, I can also do that

isyanap94 replied:
I've some issue with llama-server implementation... Somehow it can be run through my terminal, but when run the module test, it can't find llama-server command

dmb23 commented on tests/test_audio_data.py:
```diff
@@ -0,0 +1,66 @@
+        assert not output_file.exists()
+
+def test_get_audio_cache(setup_audio_loader):
+    """Test for transcript retrievals. Make sure to run this after previous test!!!"""
```
When writing tests, they should be idempotent: you can never make assumptions / requirements on the order of tests.

That means that you should restructure what is happening and create appropriate fixtures that take care of providing and cleaning up required resources.

dmb23 commented on tests/test_audio_data.py:
```diff
@@ -0,0 +1,66 @@
+    assert audio_full_texts[0].startswith("And before he had time") == True
+    assert audio_full_texts[1].startswith("And before he had time") == True
+
+    for file in audio_loader.outdir.glob("*.txt"):
```
this does not work as you hope. For me when running the full test-suite, this file is still present. Compare the comment above.

Commit:
Mostly updates on test module

Commit:
Remove unnecessary file

Commit:
Calling exe through absolute path, add some tests
